beancount_import/source/generic_importer_source.py

```diff
@@ -0,0 +1,188 @@
+
+The Transaction.narration set by each importer is copied to Posting.meta[source_desc]
+This helps in predicting postings for similar transaction while allowing the
+user to change the Transaction description and payee from UI
+(see readme.md for more on source_desc)
+This `source_desc` meta is also used for check cleared postings and should not be
+changed manually
+
+Author: Sufiyan Adhikari(github.com/dumbPy)
+"""
+
+import os
+import hashlib
+from glob import glob
+from typing import List
+from collections import defaultdict
+import itertools
+import datetime
+
+from beancount.core.data import Transaction, Posting,  Directive
+from beancount.core import data
+from beancount.core.amount import Amount
+from beancount.ingest.importer import ImporterProtocol
+from beancount.core.compare import hash_entry
+from beancount.ingest.cache import get_file
+from beancount.ingest.similar import find_similar_entries, SimilarityComparator
+
+from ..matching import FIXME_ACCOUNT, SimpleInventory
+from . import ImportResult, Source, SourceResults, InvalidSourceReference, AssociatedData
+from ..journal_editor import JournalEditor
+from .description_based_source import DescriptionBasedSource
+
+
+class ImporterSource(DescriptionBasedSource):
+    def __init__(self,
+                 directory: str,
+                 account: str,
+                 importer: ImporterProtocol,
+                 **kwargs) -> None:
+        super().__init__(**kwargs)
+        self.directory = os.path.expanduser(directory)
+        self.importer = importer
+        self.account = account
+
+        self._comparator = SimilarityComparator()
+
+        # get _FileMemo object for each file
+        files = [get_file(f) for f in
+                    filter(os.path.isfile,
+                 glob(os.path.join(directory, '**', '*'), recursive=True)
+                           )
+        ]
+        # filter the valid files for this importer
+        self.files = [f for f in files if self.importer.identify(f)]
+
+    @property
+    def name(self):
+        return self.importer.name()
+
+    def prepare(self, journal: 'JournalEditor', results: SourceResults) -> None:
+        results.add_account(self.account)
+        entries = defaultdict(list)
+        for f in self.files:
+            f_entries = self.importer.extract(f, existing_entries=journal.entries)
+            # collect  all entries in current statement, grouped by hash
+            hashed_entries = defaultdict(list)
+            for entry in f_entries:
+                hash_ = self._hash_entry(entry, frozenset(['filename','lineno']))
+                hashed_entries[hash_].append(entry)
+            # deduplicate across statements
+            for hash_ in hashed_entries:
+                # skip the existing entries from other statements. add remaining
+                n = len(entries[hash_])
+                entries[hash_].extend(hashed_entries[hash_][n:])
+
```

Comment:
You could just call get_pending_and_invalid_entries from description_based_source.py rather than using similar_entries_in_journal.  That would avoid having to duplicate similar logic here, and it would also handle invalid references.

Reply:
Done!